In [1]:
import os
os.chdir('/home/smallyan/eval_agent')

# Code Evaluation: Vector Arithmetic in Concept and Token Subspaces

## Evaluation Mode: No-Execution Code Evaluation

This notebook evaluates the code implementing circuit analysis for parallelogram arithmetic in concept and token subspaces of Llama-2-7b model activations.

### Files Evaluated:
1. **scripts/parallelograms.py** - Core helper functions
2. **scripts/all_parallelograms.py** - Main analysis script
3. **scripts/parallelogram_ranks.py** - Rank scan analysis script
4. **scripts/parallelogram_analysis.ipynb** - Visualization notebook

### Evaluation Criteria:
- **Runnable (Y/N)**: Syntactically valid, required imports available, no obvious runtime blockers
- **Correct-Implementation (Y/N)**: Logic implements described computation correctly
- **Redundant (Y/N)**: Block duplicates another block's computation
- **Irrelevant (Y/N)**: Block does not contribute to the project goal

In [2]:
import pandas as pd
import json

# Define the evaluation data structure
evaluation_data = []

# ============================================
# EVALUATION OF: scripts/parallelograms.py
# ============================================

# Function 1: logit_lens (lines 18-20)
evaluation_data.append({
    "File": "parallelograms.py",
    "Block/Function": "logit_lens",
    "Lines": "18-20",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Function 2: print_logit_lens (lines 22-24)
evaluation_data.append({
    "File": "parallelograms.py",
    "Block/Function": "print_logit_lens",
    "Lines": "22-24",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Function 3: proj_onto_ov (lines 27-42)
evaluation_data.append({
    "File": "parallelograms.py",
    "Block/Function": "proj_onto_ov",
    "Lines": "27-42",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Function 4: get_ov_sum (lines 44-74)
evaluation_data.append({
    "File": "parallelograms.py",
    "Block/Function": "get_ov_sum",
    "Lines": "44-74",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Function 5: get_neighbors (lines 77-100)
evaluation_data.append({
    "File": "parallelograms.py",
    "Block/Function": "get_neighbors",
    "Lines": "77-100",
    "Runnable": "N",
    "Correct-Implementation": "N",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": "In the else branch (lines 95-98), proj_onto_ov is called with incorrect arguments: 'k' is passed as keyword but function expects 'ov_sum' as second positional. Also missing ov_sum computation in else branch."
})

# Function 6: get_parallelogram_scores (lines 103-135)
evaluation_data.append({
    "File": "parallelograms.py",
    "Block/Function": "get_parallelogram_scores",
    "Lines": "103-135",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Function 7: all_dot_products (lines 137-195)
evaluation_data.append({
    "File": "parallelograms.py",
    "Block/Function": "all_dot_products",
    "Lines": "137-195",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "Y",
    "Error Notes": "This function is never called in any script. Not used in main analysis."
})

# Function 8: calculate_save_scores (lines 198-253)
evaluation_data.append({
    "File": "parallelograms.py",
    "Block/Function": "calculate_save_scores",
    "Lines": "198-253",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Function 9: main (lines 255-281)
evaluation_data.append({
    "File": "parallelograms.py",
    "Block/Function": "main",
    "Lines": "255-281",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

print("Evaluated: parallelograms.py (9 functions)")

Evaluated: parallelograms.py (9 functions)


In [3]:
# ============================================
# EVALUATION OF: scripts/all_parallelograms.py
# ============================================

# Function 1: loop_for_task (lines 25-40)
evaluation_data.append({
    "File": "all_parallelograms.py",
    "Block/Function": "loop_for_task",
    "Lines": "25-40",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Function 2: main (lines 42-119)
evaluation_data.append({
    "File": "all_parallelograms.py",
    "Block/Function": "main",
    "Lines": "42-119",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Block 3: __main__ block (lines 121-131)
evaluation_data.append({
    "File": "all_parallelograms.py",
    "Block/Function": "__main__",
    "Lines": "121-131",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

print("Evaluated: all_parallelograms.py (3 blocks)")

Evaluated: all_parallelograms.py (3 blocks)


In [4]:
# ============================================
# EVALUATION OF: scripts/parallelogram_ranks.py
# ============================================

# Function 1: run_rank_scan (lines 26-47)
evaluation_data.append({
    "File": "parallelogram_ranks.py",
    "Block/Function": "run_rank_scan",
    "Lines": "26-47",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Function 2: get_optimal_layers (lines 49-69)
evaluation_data.append({
    "File": "parallelogram_ranks.py",
    "Block/Function": "get_optimal_layers",
    "Lines": "49-69",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Function 3: main (lines 71-149)
evaluation_data.append({
    "File": "parallelogram_ranks.py",
    "Block/Function": "main",
    "Lines": "71-149",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Block 4: __main__ block (lines 151-161)
evaluation_data.append({
    "File": "parallelogram_ranks.py",
    "Block/Function": "__main__",
    "Lines": "151-161",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

print("Evaluated: parallelogram_ranks.py (4 blocks)")

Evaluated: parallelogram_ranks.py (4 blocks)


In [5]:
# ============================================
# EVALUATION OF: scripts/parallelogram_analysis.ipynb
# ============================================

# Cell 1: Imports and setup
evaluation_data.append({
    "File": "parallelogram_analysis.ipynb",
    "Block/Function": "Cell 1 (imports/setup)",
    "Lines": "Cell bfef1a27",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Cell 2: get_number_neighbors function
evaluation_data.append({
    "File": "parallelogram_analysis.ipynb",
    "Block/Function": "Cell 2 (get_number_neighbors)",
    "Lines": "Cell 0f373b74",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Cell 3 (markdown): Section header - skip

# Cell 4: nn_acc_word2vec function
evaluation_data.append({
    "File": "parallelogram_analysis.ipynb",
    "Block/Function": "Cell 4 (nn_acc_word2vec)",
    "Lines": "Cell e4adaa53",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Cell 5: Call nn_acc_word2vec
evaluation_data.append({
    "File": "parallelogram_analysis.ipynb",
    "Block/Function": "Cell 5 (plot word2vec)",
    "Lines": "Cell 5754b9b5",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Cell 6 (markdown): Section header - skip

# Cell 7: get_number_neighbors_fv function
evaluation_data.append({
    "File": "parallelogram_analysis.ipynb",
    "Block/Function": "Cell 7 (get_number_neighbors_fv)",
    "Lines": "Cell efa48d15",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Cell 8: nn_acc_fv function
evaluation_data.append({
    "File": "parallelogram_analysis.ipynb",
    "Block/Function": "Cell 8 (nn_acc_fv)",
    "Lines": "Cell 802bfa08",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Cell 9: Call nn_acc_fv
evaluation_data.append({
    "File": "parallelogram_analysis.ipynb",
    "Block/Function": "Cell 9 (plot fv tasks)",
    "Lines": "Cell 2e8b628e",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Cell 10 (markdown): Section header - skip

# Cell 11: single_plot function and calls
evaluation_data.append({
    "File": "parallelogram_analysis.ipynb",
    "Block/Function": "Cell 11 (single_plot)",
    "Lines": "Cell 1c59a64a",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Cell 12 (markdown): Section header - skip

# Cell 13: rank_results loading (partial)
evaluation_data.append({
    "File": "parallelogram_analysis.ipynb",
    "Block/Function": "Cell 13 (rank_results)",
    "Lines": "Cell 8964be66",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "Y",
    "Irrelevant": "N",
    "Error Notes": "This cell loads rank results but the result is not used elsewhere. The next cell reimplements the same logic more completely."
})

# Cell 14: plot_task_ranks function
evaluation_data.append({
    "File": "parallelogram_analysis.ipynb",
    "Block/Function": "Cell 14 (plot_task_ranks)",
    "Lines": "Cell 85ef187d",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

# Cell 15: Call plot_task_ranks
evaluation_data.append({
    "File": "parallelogram_analysis.ipynb",
    "Block/Function": "Cell 15 (plot ranks)",
    "Lines": "Cell 539c9f3d",
    "Runnable": "Y",
    "Correct-Implementation": "Y",
    "Redundant": "N",
    "Irrelevant": "N",
    "Error Notes": ""
})

print("Evaluated: parallelogram_analysis.ipynb (12 code cells)")

Evaluated: parallelogram_analysis.ipynb (12 code cells)


## 1.1 Block-Level Evaluation Table

In [6]:
# Create DataFrame and display the evaluation table
df = pd.DataFrame(evaluation_data)
df.index = range(1, len(df) + 1)  # 1-based indexing
df.index.name = "Block #"

# Display full table
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_rows', None)
print("=" * 120)
print("BLOCK-LEVEL EVALUATION TABLE")
print("=" * 120)
print(df.to_string())
print("\n")

BLOCK-LEVEL EVALUATION TABLE
                                 File                    Block/Function          Lines Runnable Correct-Implementation Redundant Irrelevant                                                                                                                                                                                                      Error Notes
Block #                                                                                                                                                                                                                                                                                                                                                     
1                   parallelograms.py                        logit_lens          18-20        Y                      Y         N          N                                                                                                                                      

## 1.2 Quantitative Metrics

In [7]:
# Calculate quantitative metrics
total_blocks = len(df)

# Count each flag
runnable_y = (df['Runnable'] == 'Y').sum()
runnable_n = (df['Runnable'] == 'N').sum()

correct_y = (df['Correct-Implementation'] == 'Y').sum()
correct_n = (df['Correct-Implementation'] == 'N').sum()

redundant_y = (df['Redundant'] == 'Y').sum()
redundant_n = (df['Redundant'] == 'N').sum()

irrelevant_y = (df['Irrelevant'] == 'Y').sum()
irrelevant_n = (df['Irrelevant'] == 'N').sum()

# Calculate percentages
runnable_pct = (runnable_y / total_blocks) * 100
incorrect_pct = (correct_n / total_blocks) * 100
redundant_pct = (redundant_y / total_blocks) * 100
irrelevant_pct = (irrelevant_y / total_blocks) * 100

# Calculate correction rate (blocks with potential fixes identified / blocks that failed)
failed_blocks = runnable_n + correct_n
# For the failed blocks, we have identified corrections (as described in notes):
# 1. get_neighbors - can be fixed by correcting the function call signature
identified_corrections = 1  # We identified 1 fix for the 1 failing block
correction_rate_pct = (identified_corrections / failed_blocks) * 100 if failed_blocks > 0 else 100.0

print("=" * 60)
print("QUANTITATIVE METRICS")
print("=" * 60)
print(f"\nTotal code blocks evaluated: {total_blocks}")
print(f"\nRunnable:                {runnable_y} / {total_blocks}")
print(f"Correct-Implementation:  {correct_y} / {total_blocks}")
print(f"Redundant:               {redundant_y} / {total_blocks}")
print(f"Irrelevant:              {irrelevant_y} / {total_blocks}")
print()
print("-" * 60)
print("PERCENTAGES:")
print("-" * 60)
print(f"Runnable%:               {runnable_pct:.2f}%")
print(f"Incorrect%:              {incorrect_pct:.2f}%")
print(f"Redundant%:              {redundant_pct:.2f}%")
print(f"Irrelevant%:             {irrelevant_pct:.2f}%")
print(f"Correction-Rate%:        {correction_rate_pct:.2f}%")
print("=" * 60)

QUANTITATIVE METRICS

Total code blocks evaluated: 27

Runnable:                26 / 27
Correct-Implementation:  26 / 27
Redundant:               1 / 27
Irrelevant:              1 / 27

------------------------------------------------------------
PERCENTAGES:
------------------------------------------------------------
Runnable%:               96.30%
Incorrect%:              3.70%
Redundant%:              3.70%
Irrelevant%:             3.70%
Correction-Rate%:        50.00%


## 1.3 Binary Checklist Summary

In [8]:
# Create binary checklist summary
checklist = {
    "C1": {
        "item": "All core analysis code is runnable",
        "condition": "No block has Runnable = N",
        "result": "PASS" if runnable_n == 0 else "FAIL"
    },
    "C2": {
        "item": "All implementations are correct",
        "condition": "No block has Correct-Implementation = N",
        "result": "PASS" if correct_n == 0 else "FAIL"
    },
    "C3": {
        "item": "No redundant code",
        "condition": "No block has Redundant = Y",
        "result": "PASS" if redundant_y == 0 else "FAIL"
    },
    "C4": {
        "item": "No irrelevant code",
        "condition": "No block has Irrelevant = Y",
        "result": "PASS" if irrelevant_y == 0 else "FAIL"
    }
}

print("=" * 90)
print("BINARY CHECKLIST SUMMARY")
print("=" * 90)
print(f"{'Checklist Item':<45} | {'Condition':<30} | {'PASS/FAIL':<10}")
print("-" * 90)
for key, val in checklist.items():
    print(f"{key}: {val['item']:<40} | {val['condition']:<30} | {val['result']:<10}")
print("=" * 90)

BINARY CHECKLIST SUMMARY
Checklist Item                                | Condition                      | PASS/FAIL 
------------------------------------------------------------------------------------------
C1: All core analysis code is runnable       | No block has Runnable = N      | FAIL      
C2: All implementations are correct          | No block has Correct-Implementation = N | FAIL      
C3: No redundant code                        | No block has Redundant = Y     | FAIL      
C4: No irrelevant code                       | No block has Irrelevant = Y    | FAIL      


## Summary

### Issues Found:

1. **Runnable Issue (Block #5 - get_neighbors in parallelograms.py)**:
   - In the `else` branch (lines 95-98), `proj_onto_ov` is called with incorrect arguments. The function expects `ov_sum` as the second positional argument, but it's being passed `model` as second argument and `k` as a keyword argument which doesn't exist in the function signature.
   - Additionally, `ov_sum` is not computed in the `else` branch before being used.
   - **Fix**: Compute `ov_sum = get_ov_sum(model, head_ordering, k, rank)` at the start of the `else` branch, and fix the `proj_onto_ov` calls to use the correct signature.

2. **Redundant Code (Block #25 - Cell 13 in parallelogram_analysis.ipynb)**:
   - This cell loads rank results into a variable `rank_results` that is never used. The next cell (Cell 14) reimplements this logic more completely as part of the `plot_task_ranks` function.

3. **Irrelevant Code (Block #7 - all_dot_products in parallelograms.py)**:
   - This function computes dot products and cosine similarities between word pair differences and saves histograms. However, it is never called anywhere in the codebase. According to the Plan and CodeWalkthrough, the main analysis focuses on nearest-neighbor accuracy and logit lens accuracy, not dot product histograms.

### Overall Assessment:
- The codebase is largely well-implemented with 96.30% of blocks being runnable and correctly implemented.
- One function (`get_neighbors`) has a bug in a code path that may not be exercised in the main analysis (the `else` branch for different prefixes).
- One unused function and one redundant notebook cell exist but don't affect the core analysis.

In [9]:
# Create and save JSON summary
import os

# Create output directory
output_dir = "/net/scratch2/smallyan/arithmetic_eval/no_exe_evaluation"
os.makedirs(output_dir, exist_ok=True)

# Rationale for each checklist item
rationale = {
    "C1_All_Runnable": "FAIL: Block #5 (get_neighbors in parallelograms.py) has incorrect function call arguments in the else branch (lines 95-98). The proj_onto_ov function is called with wrong parameters: 'k' is passed as keyword but function expects 'ov_sum' as second positional argument.",
    "C2_All_Correct": "FAIL: Block #5 (get_neighbors in parallelograms.py) has implementation errors. The else branch does not compute ov_sum before using it, and calls proj_onto_ov with incorrect argument signature.",
    "C3_No_Redundant": "FAIL: Block #25 (Cell 13 in parallelogram_analysis.ipynb) loads rank results into a variable that is never used. The same logic is reimplemented in the next cell.",
    "C4_No_Irrelevant": "FAIL: Block #7 (all_dot_products in parallelograms.py) is never called anywhere in the codebase and does not contribute to the main analysis goals."
}

json_summary = {
    "Runnable_Percentage": round(runnable_pct, 2),
    "Incorrect_Percentage": round(incorrect_pct, 2),
    "Redundant_Percentage": round(redundant_pct, 2),
    "Irrelevant_Percentage": round(irrelevant_pct, 2),
    "Correction_Rate_Percentage": round(correction_rate_pct, 2),
    
    "Issues": {
        "Runnable_Issues_Exist": runnable_n > 0,
        "Output_Mismatch_Exists": False,  # No execution, so no output mismatch check
        "Incorrect_Exists": correct_n > 0,
        "Redundant_Exists": redundant_y > 0,
        "Irrelevant_Exists": irrelevant_y > 0
    },
    
    "Checklist": {
        "C1_All_Runnable": checklist["C1"]["result"],
        "C2_All_Correct": checklist["C2"]["result"],
        "C3_No_Redundant": checklist["C3"]["result"],
        "C4_No_Irrelevant": checklist["C4"]["result"]
    },
    
    "Rationale": rationale
}

# Save JSON summary
json_path = os.path.join(output_dir, "code_critic_summary.json")
with open(json_path, 'w') as f:
    json.dump(json_summary, f, indent=2)

print(f"JSON summary saved to: {json_path}")
print("\nJSON Content:")
print(json.dumps(json_summary, indent=2))

TypeError: Object of type bool is not JSON serializable

In [10]:
# Convert numpy bools to Python bools
import os

# Create output directory
output_dir = "/net/scratch2/smallyan/arithmetic_eval/no_exe_evaluation"
os.makedirs(output_dir, exist_ok=True)

# Rationale for each checklist item
rationale = {
    "C1_All_Runnable": "FAIL: Block #5 (get_neighbors in parallelograms.py) has incorrect function call arguments in the else branch (lines 95-98). The proj_onto_ov function is called with wrong parameters: 'k' is passed as keyword but function expects 'ov_sum' as second positional argument.",
    "C2_All_Correct": "FAIL: Block #5 (get_neighbors in parallelograms.py) has implementation errors. The else branch does not compute ov_sum before using it, and calls proj_onto_ov with incorrect argument signature.",
    "C3_No_Redundant": "FAIL: Block #25 (Cell 13 in parallelogram_analysis.ipynb) loads rank results into a variable that is never used. The same logic is reimplemented in the next cell.",
    "C4_No_Irrelevant": "FAIL: Block #7 (all_dot_products in parallelograms.py) is never called anywhere in the codebase and does not contribute to the main analysis goals."
}

json_summary = {
    "Runnable_Percentage": float(round(runnable_pct, 2)),
    "Incorrect_Percentage": float(round(incorrect_pct, 2)),
    "Redundant_Percentage": float(round(redundant_pct, 2)),
    "Irrelevant_Percentage": float(round(irrelevant_pct, 2)),
    "Correction_Rate_Percentage": float(round(correction_rate_pct, 2)),
    
    "Issues": {
        "Runnable_Issues_Exist": bool(runnable_n > 0),
        "Output_Mismatch_Exists": False,  # No execution, so no output mismatch check
        "Incorrect_Exists": bool(correct_n > 0),
        "Redundant_Exists": bool(redundant_y > 0),
        "Irrelevant_Exists": bool(irrelevant_y > 0)
    },
    
    "Checklist": {
        "C1_All_Runnable": checklist["C1"]["result"],
        "C2_All_Correct": checklist["C2"]["result"],
        "C3_No_Redundant": checklist["C3"]["result"],
        "C4_No_Irrelevant": checklist["C4"]["result"]
    },
    
    "Rationale": rationale
}

# Save JSON summary
json_path = os.path.join(output_dir, "code_critic_summary.json")
with open(json_path, 'w') as f:
    json.dump(json_summary, f, indent=2)

print(f"JSON summary saved to: {json_path}")
print("\nJSON Content:")
print(json.dumps(json_summary, indent=2))

JSON summary saved to: /net/scratch2/smallyan/arithmetic_eval/no_exe_evaluation/code_critic_summary.json

JSON Content:
{
  "Runnable_Percentage": 96.3,
  "Incorrect_Percentage": 3.7,
  "Redundant_Percentage": 3.7,
  "Irrelevant_Percentage": 3.7,
  "Correction_Rate_Percentage": 50.0,
  "Issues": {
    "Runnable_Issues_Exist": true,
    "Output_Mismatch_Exists": false,
    "Incorrect_Exists": true,
    "Redundant_Exists": true,
    "Irrelevant_Exists": true
  },
  "Checklist": {
    "C1_All_Runnable": "FAIL",
    "C2_All_Correct": "FAIL",
    "C3_No_Redundant": "FAIL",
    "C4_No_Irrelevant": "FAIL"
  },
  "Rationale": {
    "C1_All_Runnable": "FAIL: Block #5 (get_neighbors in parallelograms.py) has incorrect function call arguments in the else branch (lines 95-98). The proj_onto_ov function is called with wrong parameters: 'k' is passed as keyword but function expects 'ov_sum' as second positional argument.",
    "C2_All_Correct": "FAIL: Block #5 (get_neighbors in parallelograms.py) ha

In [11]:
# Save evaluation table to CSV for reference
df.to_csv(os.path.join(output_dir, "block_evaluation_table.csv"))
print(f"Block evaluation table saved to: {os.path.join(output_dir, 'block_evaluation_table.csv')}")

Block evaluation table saved to: /net/scratch2/smallyan/arithmetic_eval/no_exe_evaluation/block_evaluation_table.csv
